### PR TITLE
Enhance Checkpoint Converter for MoE Models

### PR DESCRIPTION
**Pull request by parambole**
### **Title:**

**Feat: Enhance Checkpoint Converter for MoE Models**

* * * * *

### TL;DR

-   **What:** This PR enhances the checkpoint conversion framework to support MoE models with multi-axis scanned tensor layouts. It adds full, bidirectional conversion support for the **`qwen3-30b-a3b`** MoE model.

-   **Why:** To enable the conversion of MoE models.

-   **How:** By refactoring the core stacking and slicing logic in `to_maxtext.py` and `utils.py` to correctly process nested parameter mappings. This includes implementing an "experts-first" processing strategy for MoE weights and adding support for unscanned MoE layers in `param_mapping.py`.

#### Background and Motivation

The existing checkpoint conversion utility provided a strong foundation for model interoperability. However, it was unable to handle certain Mixture-of-Experts (MoE) models, like the Qwen3 series.

Specifically, while standard scanned weights in MaxText are often stored with the scan axis transposed (e.g., `(hidden_size, num_layers, ...)`), these MoE weights require a layout where the stacking axes are first: `(num_experts, num_layers, ...)`. This PR makes the framework to handle both cases correctly.

#### Key Architectural Changes

This PR introduces enhancements to the conversion framework:

1.  **Multi-Axis Stacking/Slicing Logic (`to_maxtext.py`, `utils/utils.py`):**

    -   The core conversion loops in both `to_maxtext.py` (for HF -> MaxText) and `utils.py`'s `process_leaf_param` (for MaxText -> HF) have been refactored.

    -   They can now detect **nested list mappings**, which are used to represent tensors scanned along multiple axes (experts and layers).

    -   An **"experts-first"** processing strategy has been implemented. When building a MaxText MoE tensor, the logic now iterates through experts first, then layers, correctly producing the final `(num_experts, num_layers, ...)` tensor shape without requiring an inefficient final transpose. The reverse is true when disassembling the tensor.

2.  **Unscanned MoE Support (`param_mapping.py`):**

    -   The `QWEN3_MAXTEXT_TO_HF_PARAM_MAPPING` function has been updated to fully support `scan_layers=False`.

    -   It now correctly generates mappings for unscanned MoE layers, where a single MaxText weight (e.g., `layers_0-moe_block-wi_0`) corresponds to a 1D list of all expert weights for that layer.

    -   The logic in `utils.py` was also enhanced to correctly identify and slice these unscanned MoE tensors along the expert axis.

3.  **Configuration for `qwen3-30b-a3b`:**

    -   **`utils/hf_model_configs.py`**: A new `Qwen3MoeConfig` object has been added to define the model's architecture.

    -   **`utils/hf_shape.py`**: The model key has been added to the `HF_SHAPE` dictionary, pointing to the generic `QWEN3_HF_WEIGHTS_TO_SHAPE` function.

    -   **`utils/param_mapping.py`**: The model key was registered in the `PARAM_MAPPING` and `HOOK_FNS` dictionaries. The hook assignment logic was also made more robust to handle parameter names with or without a `-kernel` suffix.

    -   **`utils/utils.py`**: The model's Hugging Face Hub ID was added to the `HF_IDS` dictionary.

4.  **Bug Fixes:**

    -   A bug in `to_maxtext.py` was fixed where the correct `target_shape` was not being passed to the hook function for standard scanned layers, causing reshape errors.

### Testing Strategy & Validation

The correctness of the implementation was validated by performing a round-trip conversion.

1.  The original Hugging Face checkpoint was converted to the MaxText format.
```
python -m MaxText.utils.ckpt_conversion.to_maxtext MaxText/configs/base.yml skip_jax_distributed_system=True scan_layers=True model_name="qwen3-30b-a3b" base_output_directory="gs://parambole-qwen3-moe-verification/scanned/qwen3-30b-a3b-thinking-2507/19_08_2025"
```
2.  The resulting MaxText checkpoint was then converted back to the Hugging Face format.
```
JAX_PLATFORMS=cpu python3 -m MaxText.utils.ckpt_conversion.to_huggingface \
  MaxText/configs/base.yml \
  model_name=qwen3-30b-a3b \
  load_parameters_path=gs://parambole-qwen3-moe-verification/scanned/qwen3-30b-a3b-thinking-2507/19_08_2025/0/items/ \
  base_output_directory="gs://parambole-qwen3-moe-verification/hf/scanned/qwen3-30b-a3b-thinking-2507/19_08_2025" \
  scan_layers=true \
  skip_jax_distributed_system=True \
  checkpoint_storage_concurrent_gb=512
```
3. Logit verification 
```
JAX_PLATFORMS=cpu python3 -m MaxText.tests.forward_pass_logit_checker MaxText/configs/base.yml normalize_embedding_logits=False logits_via_embedding=False dtype=float32 activations_in_float32=true matmul_precision=high \
max_prefill_predict_length=4 max_target_length=8 \
tokenizer_path=assets/qwen3-tokenizer \
megablox=False sparse_matmul=False scan_layers=False \
load_parameters_path=gs://parambole-qwen3-moe-verification/unscanned/qwen3-30b-a3b-thinking-2507/19_08_2025/0/items/ \
model_name=qwen3-30b-a3b \
checkpoint_storage_concurrent_gb=512 \
skip_jax_distributed_system=True \
--hf_model_path=/home/parambole/qwen3-hf/Qwen3-30B-A3B-Thinking-2507 \
--max_kl_div=0.025 \
--run_hf_model=True 
```

### Checklist

-   [X] I have performed a self-review of my code.
-   [X] I have necessary comments in my code, particularly in hard-to-understand areas.
-   [X] I have run end-to-end tests and provided the validation results above.
-   [X] I have made or will make corresponding changes to the documentation if needed.

### CODE REVIEW

**Comment by YixuanWang-99** [comment]
Thanks for improving the conversion framework and for supporting its first MoE model!

File: MaxText/utils/ckpt_conversion/to_maxtext.py [hunk 1]
```diff
@@ -149,27 +150,55 @@ def main(argv: Sequence[str]) -> None:
...
+        mt_slice_shape_list = list(mt_target_shape_final)
+        del mt_slice_shape_list[axis_to_stack]
+        mt_slice_shape = tuple(mt_slice_shape_list)
+
+        for hf_key_single in hf_source_keys_or_key:
+          if hf_key_single not in hf_state_dict_numpy:
+             raise ValueError(f"HuggingFace key {hf_key_single} (for MaxText {mt_param_key}) not found in HF state_dict.")
+          hf_tensor_numpy = hf_state_dict_numpy[hf_key_single]
+          processed_hf_tensor = apply_hook_fns(hf_tensor_numpy, mt_slice_shape, hook_fn_list_or_fn)
+          tensors_to_stack.append(processed_hf_tensor)
```
Comment (YixuanWang-99) [inline MaxText/utils/ckpt_conversion/to_maxtext.py]: What do you think about moving these repeated operations into a helper function?

Comment (hengtaoguo) [reply MaxText/utils/ckpt_conversion/to_maxtext.py]: +1. The current flow is mixed with many logics among if-else. It would be much easier to read if we decouple into smaller helper functions. Thanks!

Comment (parambole) [reply MaxText/utils/ckpt_conversion/to_maxtext.py]: Done.

**Comment by parambole** [comment]
> Thanks Param! How did the logits verification go? Could you provide a command line example in the PR description showing the round-trip conversion and logits check? Or maybe a simple decode command in MaxText.

I have updated the descriptions with the commands.

**Review by shralex — APPROVED** [review]
